fendermint/vm/interpreter/src/fvm/query.rs

```diff
@@ -99,3 +96,130 @@ where
         Ok((state, res))
     }
 }
+
+impl<DB> FvmMessageInterpreter<DB>
+where
+    DB: Blockstore + 'static + Send + Sync + Clone,
+{
+    /// Overestimation rate applied to gas to ensure that the
+    /// message goes through in the gas estimation.
+    const GAS_OVERESTIMATION_RATE: f64 = 1.25;
+    /// Default gas premium value. Inferred through a quick search through
+    /// InvokeEVM messages in filfox. The default value is only used if
+    /// the user hasn't specified a gas premium.
+    const DEFAULT_GAS_PREMIUM: u64 = 20000;
+    /// Gas search step increase used to find the optimal gas limit.
+    const GAS_SEARCH_STEP: f64 = 1.2;
+
+    fn estimate_gassed_msg(
+        &self,
+        state: &FvmQueryState<DB>,
+        msg: Box<Message>,
+    ) -> anyhow::Result<Message> {
+        let mut out = (*msg).clone();
+        // estimate the gas limit and assign it to the message
+        // do not reuse the cache
+        let ret = state.call_with_cache(*msg.clone(), false)?;
+        if ret.msg_receipt.exit_code != ExitCode::OK {
+            return Err(anyhow::anyhow!(
+                "message execution failed with error code {}",
+                ret.msg_receipt.exit_code
+            ));
```

Comment:
No, this also cannot be treated as an error and then [used with ? operator](https://github.com/consensus-shipyard/fendermint/pull/163/files#diff-5be56262407092392424776bca11de06a243db9770516d04402d1f7d6961ab6bR78). You have to return it as `Ok` and inspect the exit code, or you have to do a case analysis on the result, otherwise the process crashes.

Reply:
Fixed.